Send review notifications to service contacts

stephen-cox opened the pull request:
Adds a service contact entity that can be associated with nodes and sends notifications to service contact when content passes it's needs review date.

Fixes #12

andybroomfield commented on composer.json:
```diff
@@ -9,6 +9,7 @@
         "drupal/diff": "^1.0",
         "drupal/responsive_preview": "^2.0",
         "drupal/scheduled_transitions": "^2.1",
+        "drupal/symfony_mailer": "^1.4",
```
This might be minor as we have symfony_mailer installed, but are we sure we want to include the dependency here?

I see some new config for lando and mailhog which feels like it needs to be moved into profile or own module maybe as email config might be used in lots of places. 

It would be good to get a sense of others website email setup and if they could be supported (some might be using Gov Notifiy for their email).

I see there is a plugin for Email notifications that uses the new email system that Symfony mailer provides. Do we need to include the dependency for the plugin to work, or could it be a suggestion and dev dependency for the tests?

andybroomfield replied:
Although...
```
$ composer why symfony/mailer
drupal/core             10.2.2 requires  symfony/mailer (^6.4)    
drupal/core-recommended 10.2.2 requires  symfony/mailer (~v6.4.0) 
symfony/http-kernel     v6.4.3 conflicts symfony/mailer (<5.4)    
symfony/mime            v6.4.3 conflicts symfony/mailer (<5.4)  
```
https://www.drupal.org/node/3369935

stephen-cox replied:
The Symfony Mailer module provides a nice way to edit email template files. It's also easy to add new ones should we wish to add notifications on other events.

I can use the symfony/mailer component included with Drupal, but we would either need to hard code the email templates or implement our own templating system.

As it stands this work assumes the Drupal Symfony module is installed and provides some config for it. It should be possible to work with either, but I would need to investigate.

Do others have opinions on whether we should require this module?

Adnan-cds replied:
If drupal/core requires it then I see no issue localgov_workflows requiring it as well.

Adnan-cds replied:
> This is the Drupal Symfony Mailer module (https://www.drupal.org/project/symfony_mailer) rather than the Symfony library that is now included with Drupal core.

Got it now.  When I was testing this functionality today, apart from review reminder emails, I also received a system update reminder from the core "update" module.  That's when I realized (something Andy explains above already) the Symfony mailer module has become the site-wide mail processor.  I am trying to understand if this maybe lead to conflict if a site is already using other mail related modules such as the smtp module.

I agree with Andy when he says above "...feels like it needs to be moved into profile or own module maybe as email config might be used in lots of places."  For now, keeping Symfony mailer as a "dev" dependency of this module as per Andy's suggestion may not be a bad idea.

stephen-cox commented:
To test this PR:

Install and enable the localgov_workflows_notifications module:

```
composer require localgovdrupal/localgov_workflows:"dev-feature/1.x/12-review-emails as 1.2.3"
drush en localgov_workflows_notifications localgov_review_date
```

Create some service contacts at /admin/content/localgov-service-contact and then associate them with a node.

On the node, check the 'content reviewed' box and set the review date to yesterday.

Then run:

```
drush sset localgov_workflows_notifications.last_email_run 0
drush cron
```

Adnan-cds commented on modules/localgov_workflows_notifications/README.md:
```diff
@@ -0,0 +1,70 @@
+
+Any notifications added to the email queue will be sent at the next cron run.
+
+Needs review notifications are sent every X days for content with a needs review date that
```
"every X days" - should this wording remain as it is now?

stephen-cox replied:
>  "drupal/symfony_mailer": "^1.4",

This is the Drupal Symfony Mailer module (https://www.drupal.org/project/symfony_mailer) rather than the Symfony library that is now included with Drupal core.

> "every X days" - should this wording remain as it is now?

This number of days is configurable, but the wording can certainly be improved.

finnlewis commented:
@Adnan-cds is half way through reviewing this.... others encouraged to help review.

Adnan-cds commented:
Results from testing so far:
- None of the user roles has any of the Service contact management permissions.  Perhaps the "Editor" and "User manager" role should be given some permissions.
- I was adding a service contact at /admin/content/localgov-service-contact/add.  It either accepts a Drupal user or an external user.  The UX would be better if there were a visual clue to indicate it is *either* a Drupal user or an external user. ![image](https://github.com/localgovdrupal/localgov_workflows/assets/50206849/3c8d1aa5-0924-4374-bcbd-6d5f83b179d4)
- Under Claro, the Service contact widget is overstretched: 
![image](https://github.com/localgovdrupal/localgov_workflows/assets/50206849/701c3218-4ff0-4ae3-ae64-132000553eab)
- It would be good if the helptext of the Service contact widget in node edit forms briefly explains what it is for.
- After adding a Service contact to an **unpublished** Guide page and then setting its review date to yesterday, `drush cron` suffered a PHP error :(
```
[21-Feb-2024 16:44:38 Europe/London] TypeError: Drupal\localgov_review_date\Entity\ReviewDate::getEntity(): Return value must be of type Drupal\Core\Entity\EntityInterface, null returned in /path/to/codebase/web/modules/contrib/localgov_workflows/modules/localgov_review_date/src/Entity/ReviewDate.php on line 158 #0 /path/to/codebase/web/modules/contrib/localgov_workflows/modules/localgov_workflows_notifications/localgov_workflows_notifications.module(144): Drupal\localgov_review_date\Entity\ReviewDate->getEntity()
#1 /path/to/codebase/web/core/lib/Drupal/Core/Cron.php(335): localgov_workflows_notifications_cron()
#2 /path/to/codebase/web/core/lib/Drupal/Core/Extension/ModuleHandler.php(388): Drupal\Core\Cron->Drupal\Core\{closure}(Object(Closure), 'localgov_workfl...')
#3 /path/to/codebase/web/core/lib/Drupal/Core/Cron.php(318): Drupal\Core\Extension\ModuleHandler->invokeAllWith('cron', Object(Closure))
#4 /path/to/codebase/web/core/lib/Drupal/Core/Cron.php(159): Drupal\Core\Cron->invokeCronHandlers()
#5 /path/to/codebase/web/core/lib/Drupal/Core/ProxyClass/Cron.php(75): Drupal\Core\Cron->run()
#6 /path/to/codebase/vendor/drush/drush/src/Commands/core/DrupalCommands.php(63): Drupal\Core\ProxyClass\Cron->run()
#7 [internal function]: Drush\Commands\core\DrupalCommands->cron(Array)
#8 /path/to/codebase/vendor/consolidation/annotated-command/src/CommandProcessor.php(276): call_user_func_array(Array, Array)
#9 /path/to/codebase/vendor/consolidation/annotated-command/src/CommandProcessor.php(212): Consolidation\AnnotatedCommand\CommandProcessor->runCommandCallback(Array, Object(Consolidation\AnnotatedCommand\CommandData))
#10 /path/to/codebase/vendor/consolidation/annotated-command/src/CommandProcessor.php(175): Consolidation\AnnotatedCommand\CommandProcessor->validateRunAndAlter(Array, Array, Object(Consolidation\AnnotatedCommand\CommandData))
#11 /path/to/codebase/vendor/consolidation/annotated-command/src/AnnotatedCommand.php(387): Consolidation\AnnotatedCommand\CommandProcessor->process(Object(Symfony\Component\Console\Output\ConsoleOutput), Array, Array, Object(Consolidation\AnnotatedCommand\CommandData))
#12 /path/to/codebase/vendor/symfony/console/Command/Command.php(326): Consolidation\AnnotatedCommand\AnnotatedCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /path/to/codebase/vendor/symfony/console/Application.php(1081): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 /path/to/codebase/vendor/symfony/console/Application.php(320): Symfony\Component\Console\Application->doRunCommand(Object(Consolidation\AnnotatedCommand\AnnotatedCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /path/to/codebase/vendor/symfony/console/Application.php(174): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /path/to/codebase/vendor/drush/drush/src/Runtime/Runtime.php(110): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /path/to/codebase/vendor/drush/drush/src/Runtime/Runtime.php(40): Drush\Runtime\Runtime->doRun(Array, Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /path/to/codebase/vendor/drush/drush/drush.php(139): Drush\Runtime\Runtime->run(Array)
#19 /path/to/codebase/vendor/drush/drush/drush(4): require('/path/to/codebase/v...')
#20 /path/to/codebase/bin/drush(120): include('/path/to/codebase/v...')
#21 {main}
TypeError: Drupal\localgov_review_date\Entity\ReviewDate::getEntity(): Return value must be of type Drupal\Core\Entity\EntityInterface, null returned in Drupal\localgov_review_date\Entity\ReviewDate->getEntity() (line 158 of /path/to/codebase/web/modules/contrib/localgov_workflows/modules/localgov_review_date/src/Entity/ReviewDate.php).
```

ekes commented on modules/localgov_workflows_notifications/localgov_workflows_notifications.module:
```diff
@@ -0,0 +1,152 @@
+      '#description' => t('Search for a service contact by name or email.'),
+      '#group' => 'advanced',
+      '#open' => TRUE,
+      '#weight' => -4,
```
How is this going to play if people have altered the form already? Thinking the weight mainly, could it end up hard coded somewhere unintended? Or as it's in advanced is it just maybe irritating, but not important?

stephen-cox replied:
Not sure now many people form alter the advanced setting, but the weight is there to move is below the review date widget which has a weight of -5 https://github.com/localgovdrupal/localgov_workflows/blob/2f1424ef47e64b7ff3ee04c3a25577d204b4200d/modules/localgov_review_date/localgov_review_date.module#L26

ekes commented on modules/localgov_workflows_notifications/localgov_workflows_notifications.module:
```diff
@@ -0,0 +1,152 @@
+  foreach ($localgov_service_contacts as $value) {
+    if (is_array($value) && !empty($value['target_id'])) {
+      if (in_array($value['target_id'], $values)) {
+        $form_state->setError($form['localgov_service_contacts'], t('Service contacts must be unique.'));
```
A cursory scan suggests that the ConstraintValidators for a field don't include a Unique within the field list validator. Or is that because the Symfony one does the trick?

Anyway it would be nicer to ->setPropertyConstraint on the field itself and use a validator?

stephen-cox replied:
@ekes It's not possible to use the unique constraint (`addConstraint('UniqueField')`) on the field because this means you an only associate 1 service contact with 1 node.

I've been looking for examples of how you might use `addPropertyConstraints` \ `setPropertyConstraints` with a callback to say what a field should not be, as opposed to what is should be, and I can't find any. Are you aware of how this might work?

ekes reviewed (commented):
Read the code, not run it, but from that with a couple of minor points, happy.

stephen-cox commented:
Thanks for looking @Adnan-cds

> I would have just used Drupal's user accounts as Service contacts instead of adding a new Entity type

This came out of the discussion early on. Some councils have all their users in the site, some don't. This should cover both setups.

> None of the user roles has any of the Service contact management permissions. Perhaps the "Editor" and "User manager" role should be given some permissions.

Good catch. Will add permissions. Certainly to the editor role. Not sure about the user manager, as this user currently only has permissions to create users with certain roles.

Will need an install hook to add the permissions to editors.

> The UX would be better if there were a visual clue to indicate it is either a Drupal user or an external user. 

Do you have anything in mind for this? I have added some text to explain it and disabled the other fields once text is added to one field, but not sure what else can be added to make this clear.

I'll address the bugs this morning.

stephen-cox commented:
@Adnan-cds 

I have added a hook to add permissions to view and manage service contacts. Editors have permissions to manage them. Authors and contributors have permissions to view them so they can be added and removed from content.

> Under Claro, the Service contact widget is overstretched

I'm not seeing this on a fresh LGD install. Can you provide more information on your install?

![image](https://github.com/localgovdrupal/localgov_workflows/assets/7189914/0b4f6567-3599-4e28-95e3-ef3c5cce02bd)

> It would be good if the helptext of the Service contact widget in node edit forms briefly explains what it is for.

I've updated the message to 'Associate service contacts with this content by searching by name or email.'

> After adding a Service contact to an unpublished Guide page and then setting its review date to yesterday, drush cron suffered a PHP error :(

Unfortunately, I'm not seeing this either with a fresh install and the demo module. I assume you're testing this on an existing site.

The error message indicates an issue with the Review Date entity. If there's no entity associated with the review date then the getEntity command fails as it returns NULL. As we're checking for this anyway, I have updated the code to allow NULL returns, which should fix the error you're seeing.

Can you test again with the latest changes?

Adnan-cds commented:
> I have added a hook to add permissions to view and manage service contacts. Editors have permissions to manage them. Authors and contributors have permissions to view them so they can be added and removed from content.

Thanks.  Looks good.

>> Under Claro, the Service contact widget is overstretched
> I'm not seeing this on a fresh LGD install. Can you provide more information on your install?

This is no longer happening.  Thanks again :)


>> It would be good if the helptext of the Service contact widget in node edit forms briefly explains what it is for.
> I've updated the message to 'Associate service contacts with this content by searching by name or email.'

Thanks.  It still doesn't fully hint at what these Service contacts are for.  Adding another sentence such as "They will receive review reminders" may bring more clarity.

>> After adding a Service contact to an unpublished Guide page and then setting its review date to yesterday, drush cron suffered a PHP error :(
> The error message indicates an issue with the Review Date entity. If there's no entity associated with the review date then the getEntity command fails as it returns NULL. As we're checking for this anyway, I have updated the code to allow NULL returns, which should fix the error you're seeing.

No fatal error any more.  Thanks a lot :)

> Can you test again with the latest changes?

It works!

Adnan-cds commented:
>> The UX would be better if there were a visual clue to indicate it is either a Drupal user or an external user.
> Do you have anything in mind for this?

Something like the following would be good.  But this is not a must-have.  Once the question of whether or not to bundle Symfony mailer with this change is resolved, I am happy for it to be merged.
![screenshot-service-contact-add-form-ux-improvement-proposal](https://github.com/localgovdrupal/localgov_workflows/assets/50206849/1a1eaf3f-80a9-4edc-a578-f99397545180)

finnlewis commented:
Sounds like https://www.drupal.org/project/symfony_mailer is the way... and allows people full flexibility over how they configure the transport.

I will test and have a think reflecting on the comments above. 

Questions: 

1. do we force people to use Symfony Mailer? 
2. If so, do we include it as a part of the localgov profile or localgov_core for consistency? 
3. Do we write some code to accommodate people who are using other emailing solutions?

finnlewis commented:
Worth testing with 
- https://www.drupal.org/project/smtp 
- https://www.drupal.org/project/mailsystem

Adnan-cds commented:
> Worth testing with

Installed the new localgov_workflows_notifications submodule on an existing site that's been using the smtp module and symfony_mailer has indeed taken over from the smtp module.

On the upside, the Webform module [whitelists symfony_mailer](https://git.drupalcode.org/project/webform/-/blob/6.2.x/src/Plugin/WebformHandler/EmailWebformHandler.php?ref_type=heads#L1445) for sending out attachments.  So there should not be any loss of functionality for Webforms at least.

finnlewis commented:
Can we add a check / warning with 

 - hook_install
 - hook_requirements
 - incompatible modules ?

Such that: if any of these modules are installed: (smtp / mailsystem / phpmailer) - warn the user that we will install symfony_mailer and we are taking over.

Or, notify the user (developer) on install, regarding of the other modules installed?

stephen-cox commented:
@finnlewis @Adnan-cds There's now a warning message if enabling this and a conflicting module is already enabled:

https://github.com/localgovdrupal/localgov_workflows/blob/f7b720b067db7576f8c43ee0fc1743c422573cd6/modules/localgov_workflows_notifications/localgov_workflows_notifications.install#L13-L25

And a warning added to the top of the README

https://github.com/localgovdrupal/localgov_workflows/blob/e6d90c1413e2ddc443ef0a9a07de03c5453c050b/modules/localgov_workflows_notifications/README.md?plain=1#L7-L17

Do you think this sufficient warning about the potential email changes?

Adnan-cds commented:
> Do you think this sufficient warning about the potential email changes?

I can live with this.  Thanks for this.  I will test again and see how it goes.

Adnan-cds reviewed (approved):
Tested the updated hook_install() implementation and it warned me about the existing smtp module.  I am happy with the messaging.
